src/Features/Core/Portable/Diagnostics/DiagnosticService.cs

```diff
@@ -314,17 +322,35 @@ private void AssertIfNull(ImmutableArray<DiagnosticData> diagnostics)
         }
 
         [Conditional("DEBUG")]
-        private void AssertIfNull(DiagnosticData diagnostic)
+        private void AssertIfNull<T>(T obj) where T : class
         {
-            if (diagnostic == null)
+            if (obj == null)
             {
                 Contract.Requires(false, "who returns invalid data?");
             }
         }
 
+        [Conditional("DEBUG")]
+        public void AssertCleanup()
+        {
+            foreach (var idAndData in _map.Values.ToList())
+            {
+                foreach (var data in idAndData.Values.ToList())
+                {
+                    AssertIfNull(data.Workspace);
+                }
+            }
+        }
+
         private struct Data : IEquatable<Data>
         {
-            public readonly Workspace Workspace;
+            private static readonly ConditionalWeakTable<Workspace, WeakReference<Workspace>> s_cache =
+                new ConditionalWeakTable<Workspace, WeakReference<Workspace>>();
+
+            // not all workspace lives whole VS session. 
+            // make sure we not strongly hold onto them.
```

Comment:
Why is this necessary?  What keeps the "Data" alive?  Why does it go away when the workspace goes away.

This pattern concerns me because we have data living longer than the items they should be associated with.